bdemeta, cmake, resolver: support for conan packages

RichardRohac opened the pull request:
Adds optional 'conan_roots' configuration entry which overlays all the
roots that will be sourced from conan packages.

frutiger commented on bdemeta/__main__.py:
```diff
@@ -74,6 +74,21 @@ def get_parser() -> argparse.ArgumentParser:
+            raise InvalidPathError(root)
+
+    return result
+
```
Thanks for pulling this out into its own function. It would be a good idea to move it into `bdemeta/resolver.py` and write a separate unit test for it.

RichardRohac replied:
Done.

frutiger commented on bdemeta/__main__.py:
```diff
@@ -85,18 +100,9 @@ def make_resolver(config_path_str: str,
-            raise InvalidPathError(root)
+    config['roots'] = normalize_roots(config['roots'], config_dir)
+    if 'conan_roots' in config:
+        config['conan_roots'] = normalize_roots(config['conan_roots'], config_dir)
```
Based on my understanding of the change in `bdemeta/resolver.py`, it looks like the logic is: "if we identify this name as a valid target but its root is in `conan_roots`, it should be provided by a target named `CONAN_PKG::<name>`".

Would it be simpler implementation-wise (though more verbose configuration-wise) to have a `conan_targets` configuration key where we list all the targets that should come from Conan?

frutiger replied:
> Would it be simpler implementation-wise (though more verbose configuration-wise)...

Just to be clear, I'm not sure if the above trade-off is a good one, but I wanted to hear your thoughts.

RichardRohac replied:
It seems to me that this approach would be worth the maintenance burden if only some of the root's targets were provided by Conan.

Currently, there may be need to patch a given root so that it is compatible with bdemeta. I reckon this can be avoided if Conan provides all the root's targets, making the root "transparent" to bdemeta.

frutiger commented on bdemeta/resolver.py:
```diff
@@ -156,23 +157,36 @@ def _is_cmake(root: Path, name: str) -> Optional[Path]:
+
+            if root in self._conan_roots:
+                # This target will be provided by conan.
+                return Identification('conan', root)
```
If we supported an explicit `conan_targets` property, this implementation could go back to its original simpler kind and just return `Identification('conan', path)`. It also looks like the `root` here is unused, right?

RichardRohac replied:
`root` is used to select correct target, see line 267.

frutiger reviewed (commented):
Thanks for the change, I like the overall direction though I have some questions I have left in-line.

In addition to those, I have a general question: if a top-level target is in a Conan root, we would not generate any target for it in `cmake` as I understand it since [the existing generator](https://github.com/frutiger/bdemeta/blob/bef1d3612169ca8445a51420316d020830da24c8/bdemeta/cmake.py#L234-L244) does not generate anything if the type is merely a `Target`. What should we do in this scenario?

Also, before we land this change, please be sure to include an update to the `README.md` to describe this behaviour and to bump the minor version in `setup.py`.

frutiger commented:
Sorry for the slow response on this. I was wondering if there is a way to detect if a root is a conan root? Maybe the presence of a recipe `conanfile.py`?

RichardRohac commented:
> Sorry for the slow response on this. I was wondering if there is a way to detect if a root is a conan root? Maybe the presence of a recipe `conanfile.py`?

Recipes do not necessarily need to be located inside roots so that would require assumptions about the directory structure on bdemeta's side.

frutiger reviewed (approved):
Sorry for the long delay on this but it looks good to me, thanks again for the PR. 

Unfortunately [coverage has dropped](https://app.codecov.io/gh/frutiger/bdemeta/compare/27/changes#D1L284) (you can test this locally using the following command: `coverage run -m unittest && coverage html && open htmlcov/index.html`). Do you want to fix the coverage as part of this PR or separately? I'm happy either way.

RichardRohac commented:
> Sorry for the long delay on this but it looks good to me, thanks again for the PR.
> 
> Unfortunately [coverage has dropped](https://app.codecov.io/gh/frutiger/bdemeta/compare/27/changes#D1L284) (you can test this locally using the following command: `coverage run -m unittest && coverage html && open htmlcov/index.html`). Do you want to fix the coverage as part of this PR or separately? I'm happy either way.

Windows unit test needs fixing too so I will do that as well as fixing the coverage.

Commit:
bdemeta, cmake, resolver: support for conan packages

Adds optional 'conan_roots' configuration entry which lists all the
roots that will be sourced from conan packages.

frutiger commented:
Looks good. Please update the `README.md` describing this change and `setup.py` bumping its minor version.

frutiger reviewed (approved):
🚀
